awx/main/scheduler/task_manager.py

```diff
@@ -510,9 +486,39 @@ def process_pending_tasks(self, pending_tasks):
                 self.start_task(task, None, task.get_jobs_fail_chain(), None)
                 continue
 
+            # Determine if ther is control capacity for the task
+            control_instance = InstanceGroup.fit_task_to_most_remaining_capacity_instance(
+                task, self.graph['controlplane']['instances'], impact=settings.AWX_CONTROL_NODE_TASK_IMPACT, capacity_type='control'
+            ) or InstanceGroup.find_largest_idle_instance(self.graph['controlplane']['instances'], capacity_type='control')
+            if not control_instance:
+                logger.debug(f"Skipping task {task.log_format} in pending, not enough capacity left on controlplane to control new tasks")
+                continue
+
+            task.controller_node = control_instance.hostname
+
+            # All task.capacity_type == 'control' jobs should run on control plane, no need to loop over instance groups
+            if task.capacity_type == 'control':
+                # This is a task that actually needs to run in the controlplane
+                is_idle = control_instance.jobs_running == 0
+                has_capacity_for_task_impact_and_control = control_instance.remaining_capacity >= (task.task_impact + settings.AWX_CONTROL_NODE_TASK_IMPACT)
+                if not has_capacity_for_task_impact_and_control and not is_idle:
+                    # As in other places, we accept an idle instance if the node with most capacity (the control node we already selected)
+                    # does not have enough capacity.
+                    logger.debug(f"Not enough control capacity on {control_instance} to run {task.log_format}")
+                    break
```

Comment:
@AlanCoding I think I should acutally call `continue` here and add
```
                task.log_lifecycle("needs_capacity")
                job_explanation = gettext_noop("This job is not ready to start because there is not enough available control capacity.")
```
what do you think of that

Reply:
I think you are correct, otherwise we're ejecting too early out of the `for task in pending_tasks` loop.

Reply:
I don't think you need any of L501-L508. What question is this asking that wasn't already asked above when `control_instance` was set?

If the control plane doesn't have capacity for the control action (rewording your variable name), then `control_instance` will be `None` and you will have hit the prior `continue`. If all nodes had less than 5 capacity (however unlikely) but there were idle nodes, then the logic to define `control_instance` handles that too.

Reply:
basically this catches the case when a node has 5 capacity but not 6 :laughing: maybe not worth it

Reply:
Well, it wouldn't be worth it, but there's a better way to do that

```python
            control_instance = InstanceGroup.fit_task_to_most_remaining_capacity_instance(
                task, self.graph['controlplane']['instances'], impact=settings.AWX_CONTROL_NODE_TASK_IMPACT, capacity_type='control'
            ) or InstanceGroup.find_largest_idle_instance(self.graph['controlplane']['instances'], capacity_type='control')
```

You are specifically managing the impact here, `impact=settings.AWX_CONTROL_NODE_TASK_IMPACT`. If you have a "control" type task, just add 1 to that impact number you pass in.

Reply:
we addressed this in the patch you helped me with